### PR TITLE
Fix: #2100 Building shared libraries with version suffix, along with the symbolic link

### LINKED ISSUE #2100: [Build] Building shared libraries with version suffix, along with the symbolic link
A recommended practice for the shared libraries deployed in the unix-like systems is have the libraries suffixed with version, along with the symbolic links pointing to them.
Something like - 

```
libopentelemetry_trace.so  -> libopentelemetry_trace.so.1 -> libopentelemetry_trace.so.1.9.0
```

It would be good to update the CMake build to generate them. This could be a build breaking change - as libopentelemetry_trace.so would now become the sym link, and not the actual file. so this could be introduced under separate cmake option.

This issue is specifically for CMake. The bazel build change can be introduced with separate issue.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #2100

## Changes

Versioning policy is described here for debian - https://www.debian.org/doc/debian-policy/ch-sharedlibs.html. In general same is followed across other *unix distributoins. 
- Added CMake option OTELCPP_VERSIONED_LIBS to enable generating versioned shared libs.
- Currently using OPENTELEMETRY_ABI_VERSION_NO as SOVERSION, we can also use MAJOR version. But as per the recommendation, SOVERSION should change for any ABI breaking change.
- The output is as below:

```console
$ ls -ltr /usr/local/lib/libopentelemetry_*common*so*
-rw-r--r-- 1 root root 281640 Apr 20 21:59 /usr/local/lib/libopentelemetry_common.so.1.9.0
lrwxrwxrwx 1 root root     32 Apr 20 22:01 /usr/local/lib/libopentelemetry_common.so.1 -> libopentelemetry_common.so.1.9.0
lrwxrwxrwx 1 root root     28 Apr 20 22:01 /usr/local/lib/libopentelemetry_common.so -> libopentelemetry_common.so.1

$ objdump -p /usr/local/lib/libopentelemetry_common.so.1.9.0 | grep SONAME
  SONAME               libopentelemetry_common.so.1
$
```

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -188,6 +188,9 @@ option(OTELCPP_MAINTAINER_MODE "Build in maintainer mode (-Wall -Werror)" OFF)
 
 option(WITH_OPENTRACING "Whether to include the Opentracing shim" OFF)
 
+option(OTELCPP_VERSIONED_LIBS "Whether to generate the versioned shared libs"
+       OFF)
+
```
Comment (marcalff) [inline CMakeLists.txt:193]: This option only applies for builds using shared libraries.

Suggestion for the "Verify options dependencies" section in CMakeLists.txt below.

```
if(OTELCPP_VERSIONED_LIBS AND NOT BUILD_SHARED_LIBS)
  message(FATAL_ERROR "OTELCPP_VERSIONED_LIBS=ON requires BUILD_SHARED_LIBS=ON")
endif()
```

**Review by marcalff — COMMENTED** [review]
Using OPENTELEMETRY_ABI_VERSION_NO as SOVERSION looks like the correct choice.

See minor comment on dependent options.

File: CMakeLists.txt (hunk 1 above)
Comment (lalitb) [reply CMakeLists.txt:193]: Good point. Have added this check.

**Review by marcalff — APPROVED** [review]
LGTM.

To clarify, I don't think opentelemetry-cpp is ready to use shared libraries _yet_.

We need to make sure the proper symbol is part of the proper library, see issue #1429.

This is however a different issue, which should not prevent this part to go in the main branch.

Ok to merge.

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**